libbpf-rs/src/object.rs

```diff
@@ -7,57 +7,126 @@ use crate::*;
     }
 
-    /// Acquire ownership of [`Map`] s in this object for which `f` returns `true`.
-    pub fn take_maps<F>(&mut self, _f: F) -> Vec<Map>
-    where
-        F: FnMut(&Map),
-    {
+    pub fn get_name(&self) -> &str {
         unimplemented!();
     }
 
-    pub fn progs(&self) -> Vec<&Program> {
+    pub fn get_map<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut Map> {
         unimplemented!();
     }
 
-    /// Acquire ownership of [`Program`]s in this object for which `f` returns `true`.
-    pub fn take_progs<F>(&mut self, _f: F) -> Vec<Program>
-    where
-        F: FnMut(&Program),
-    {
+    pub fn get_prog<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut Program> {
         unimplemented!();
     }
 }
 
-/// Options to configure [`Object`] processing.
-pub struct ObjectOptions {}
+/// Represents a parsed but not yet loaded map.
+pub struct Map {}
+
+impl Map {
+    pub fn name<T: AsRef<str>>(&mut self, _name: T) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn map_type(&mut self, _map_type: MapType) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn key_size(&mut self, _size: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn value_size(&mut self, _size: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn max_entries(&mut self, _entries: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn numa_node(&mut self, _node: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn btf_fd(&mut self, _fd: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn btf_key_type_id(&mut self, _id: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn btf_value_type_id(&mut self, _id: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn map_ifindex(&mut self, _idx: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn inner_map_fd(&mut self, _fd: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn btf_vmlinux_value_type_id(&mut self, _id: u32) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn no_prealloc(&mut self) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn no_common_lru(&mut self) -> &mut Self {
+        unimplemented!();
+    }
+
+    // TODO: more flags here:
+    // https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L340-L341
+
+    pub fn load(&mut self) -> Result<LoadedMap> {
+        unimplemented!();
+    }
+}
 
-/// Represents a map.
+/// Represents a created map.
+///
+/// The kernel ensure the atomicity and safety of operations on a `LoadedMap`. Therefore,
+/// this handle is safe to clone and pass around between threads. This is essentially a
+/// file descriptor.
 ///
 /// Some methods require working with raw bytes. You may find libraries such as
 /// [`plain`](https://crates.io/crates/plain) helpful.
-pub struct Map {}
+#[derive(Clone)]
+pub struct LoadedMap {}
 
-impl Map {
+impl LoadedMap {
```

Comment:
I'd imagine that LoadedMap and LoadedProg would be the most common types that user code has to interact with, so I'd try to keep the name as short and and simple, as possible: Map and Prog. What you have as Map and Program right now, it's more of a MapBuilder and ProgramBuilder, no?

Reply:
ok